prometheus/metric.go

```diff
@@ -142,3 +145,31 @@ func NewInvalidMetric(desc *Desc, err error) Metric {
 func (m *invalidMetric) Desc() *Desc { return m.desc }
 
 func (m *invalidMetric) Write(*dto.Metric) error { return m.err }
+
+type timestampedMetric struct {
+	Metric
+	t time.Time
+}
+
+func (m timestampedMetric) Write(pb *dto.Metric) error {
+	e := m.Metric.Write(pb)
+	pb.TimestampMs = proto.Int64(m.t.Unix()*1000 + int64(m.t.Nanosecond()/1000000))
+	return e
+}
+
+// NewMetricWithTimestamp returns a new Metric wrapping the provided Metric in a
```

Comment:
This sounds like it works with direct instrumenation, I think we should try to limit this to MustNewConstMetric. Everyone I've seen trying to do timestamps on direct instrumentation was otherwise using the library incorrectly.

Reply:
The usual use case is already described in the doc comment. Artificially limiting this to the output of `MustNewConstMetric` (and `MustNewConstSummary` and `MustNewConstHistogram` and `NewConstMetric` and `NewConstSummary` and `NewConstHistogram`) would be weird and cumbersome. Using this wrapper is arcane enough to not happen accidentally. Whoever elects to use such a contraption without reading the doc comments is on their own.

Reply:
Additional point: Since the timestamp is immutable, people that incorrectly wrap metrics from direct instrumentation will quickly notice that they are exposing the same timestamp forever.

Reply:
That's true.